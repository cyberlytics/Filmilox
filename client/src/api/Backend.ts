export default class Backend {
    static fetchData = async () => {
        return [];
    };
}
