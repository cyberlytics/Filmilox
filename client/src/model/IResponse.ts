export interface IRegisterResponse {
    status: boolean;
}

export interface ILoginResponse {
    token: string;
    user: {
        username: string;
        email: string;
    };
}
