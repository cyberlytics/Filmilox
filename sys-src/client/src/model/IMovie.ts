export interface IMovie {
    title: string;
    description: string;
    trailer: string;
    release: Date | null;
    image?: string;
    rating?: number;
}
