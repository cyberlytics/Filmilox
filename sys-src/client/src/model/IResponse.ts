export interface IRegisterResponse {
    status: boolean;
}

export interface ILoginResponse {
    token: string;
}
