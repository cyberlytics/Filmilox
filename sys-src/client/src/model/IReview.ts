export interface IReview {
    movieId: string;
    rating: number;
    comment: string;
}
