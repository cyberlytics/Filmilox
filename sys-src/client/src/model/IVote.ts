export interface IVote {
    upvote: number;
    downvote: number;
}
