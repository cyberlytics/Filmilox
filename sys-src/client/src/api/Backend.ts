import Axios from 'axios';
import ApiRouter from './ApiRouter';
import {
    ILoginResponse,
    IMovieResponse,
    IRegisterResponse,
} from '../model/IResponse';
import { IRegister } from '../model/IRegister';
import { ILogin } from '../model/ILogin';
import { IReviewAdd, IReviewGet } from '../model/IReview';
import { IMovie, IMovieWithID } from '../model/IMovie';

export default class Backend {
    static register = async ({ email, username, password }: IRegister) => {
        try {
            const {
                data: { status },
            } = await Axios.post<IRegisterResponse>(ApiRouter.Register, {
                email,
                username,
                password,
            });
            return status;
        } catch (e) {
            throw e;
        }
    };

    static login = async ({ password, identifier }: ILogin) => {
        try {
            const {
                data: { token },
            } = await Axios.post<ILoginResponse>(ApiRouter.Login, {
                password,
                identifier,
            });
            localStorage.setItem('token', token);
        } catch (e) {
            throw e;
        }
    };

    static addreview = async ({ movieId, rating, comment }: IReviewAdd) => {
        try {
            const { data } = await Axios.post(
                ApiRouter.AddReview,
                { movieId, rating, comment },
                ApiRouter.createHeaders()
            );
            return data;
        } catch (e) {
            throw e;
        }
    };

    static getReview = async ({ movieId }: { movieId: string }) => {
        try {
            const url = `${ApiRouter.GetReview}/${movieId}`;
            const { data } = await Axios.get(url);
            return data;
        } catch (e) {
            throw e;
        }
    };

    static getUsername = async ({ userId }: { userId: string }) => {
        try {
            const url = `${ApiRouter.GetUsername}/${userId}`;
            const { data } = await Axios.get(url);
            return data.username;
        } catch (e) {
            throw e;
        }
    };

    static addMovie = async (formData: FormData) => {
        try {
            const {
                data: { status },
            } = await Axios.post<IMovieResponse>(
                ApiRouter.AddMovie,
                formData,
                ApiRouter.createHeaders()
            );
            return status;
        } catch (e) {
            throw e;
        }
    };
    static getMovie = async ({ _id }: { _id: string }) => {
        try {
            const url = `${ApiRouter.GetMovieData}/${_id}`;
            const { data } = await Axios.get<IMovie>(url);
            return data;
        } catch (e) {
            throw e;
        }
    };

    static search = async (query: string) => {
        try {
            const { data } = await Axios.get<Array<IMovieWithID>>(
                `${ApiRouter.Search}?q=${query}`
            );
            return data;
        } catch (e) {
            throw e;
        }
    };
}
