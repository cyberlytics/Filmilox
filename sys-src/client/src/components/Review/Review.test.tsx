import Review from './Review';
import { IMovie } from '../../model/IMovie';
import { IReviewGet } from '../../model/IReview';
import { useState } from 'react';
import { render } from '../../utils/test-utils';
import Backend from '../../api/Backend';
import { waitFor } from '@testing-library/react';

const movieTest: IMovie = {
    _id: 'random_id',
    title: 'Titel',
    description: 'Beschreibung',
    trailer: 'Trailer-Link',
    release: new Date(),
    image: '/cover.png',
    rating: 2,
};
const reviewTest: IReviewGet = {
    _id: 'test_id',
    user: { _id: 'test_id', username: 'TestUser' },
    movie: 'random_id',
    rating: 2,
    comment: 'Beschreibung',
    createdAt: new Date().toString(),
    updatedAt: new Date().toString(),
};

jest.mock('../../api/Backend');

const Wrapper = () => {
    const [movie, setMovie] = useState<IMovie>(movieTest);
    const [reviews, setReviews] = useState<IReviewGet[]>([]);
    return (
        <Review
            setMovie={setMovie}
            setReviews={setReviews}
            review={reviewTest}
        />
    );
};

describe('Test Voting on Review', () => {
    it('should render the placeholders if no votes are found', async () => {
        jest.spyOn(Backend, 'getVotes').mockResolvedValue({
            data: {
                upvote: 0,
                downvote: 0,
            },
            status: 200,
            statusText: 'OK',
            headers: {},
            config: {},
        });
        const { getByTestId } = render(<Wrapper />);

        await waitFor(() => {
            const downvotePlaceholder = getByTestId(
                'downvote-count-placeholder'
            );
            const upvotePlaceholder = getByTestId('upvote-count-placeholder');
            expect(downvotePlaceholder).toBeInTheDocument();
            expect(upvotePlaceholder).toBeInTheDocument();
        });
    });
    it('should display the vote counts', async () => {
        jest.spyOn(Backend, 'getVotes').mockResolvedValue({
            data: {
                upvote: 100,
                downvote: 50,
            },
            status: 200,
            statusText: 'OK',
            headers: {},
            config: {},
        });
        const { getByTestId } = render(<Wrapper />);
        await waitFor(() => {
            const downvoteCount = getByTestId('downvote-count');
            const upvoteCount = getByTestId('upvote-count');
            expect(downvoteCount).toBeInTheDocument();
            expect(downvoteCount).toHaveTextContent('50');
            expect(upvoteCount).toBeInTheDocument();
            expect(upvoteCount).toHaveTextContent('100');
        });
    });
});
//////////////////////////////////////////////////////////////////////////////////////////////
describe('Test MovieCard Component', () => {
    test('check UI elements', () => {
        const { getByTestId } = render(<Wrapper />);

        //check if main Review div is rendered
        const mainDiv = getByTestId('review-main');
        expect(mainDiv).toBeInTheDocument();

        //check Review card div
        const cardDiv = getByTestId('review-card');
        expect(cardDiv).toBeInTheDocument();

        //check review authors username
        const username = getByTestId('addreview-username');
        expect(username).toBeInTheDocument();

        //check rating in the review
        const rating = getByTestId('addreview-rating');
        expect(rating).toBeInTheDocument();

        //check date of the review
        const reviewDate = getByTestId('addreview-date');
        expect(reviewDate).toBeInTheDocument();

        //check delete icon of the review
        const deleteIcon = getByTestId('deletereview-button');
        expect(deleteIcon).toBeInTheDocument();

        //check text the review
        const reviewText = getByTestId('addreview-comment');
        expect(reviewText).toBeInTheDocument();
    });
});
