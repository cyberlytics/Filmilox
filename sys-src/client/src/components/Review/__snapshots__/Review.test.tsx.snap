// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Review component will check the date 1`] = `
Object {
  "createdAt": Any<Date>,
}
`;
